src/GraphQL.php

```diff
@@ -373,4 +390,54 @@ protected function parseQuery(string $query): DocumentNode
             'noLocation' => ! $this->configRepository->get('lighthouse.parse_source_location'),
         ]);
     }
+
+    /**
+     * Provides a result for cacheable validation rules by running them or retrieving it from the cache.
+     *
+     * @param  array<string, \GraphQL\Validator\Rules\ValidationRule>  $validationRules
+     *
+     * @return array<Error>
+     */
+    protected function validateCacheableRules(
+        array $validationRules,
+        Schema $schema,
+        string $schemaHash,
+        DocumentNode $query,
+        ?string $queryHash,
+    ): array {
+        foreach ($validationRules as $rule) {
+            if ($rule instanceof QueryComplexity) {
+                throw new \InvalidArgumentException('The QueryComplexity rule must not be registered in cacheableValidationRules, as it depends on variables.');
+            }
+        }
+
+        if ($queryHash === null) {
+            return DocumentValidator::validate($schema, $query, $validationRules);
+        }
+
+        $cacheConfig = $this->configRepository->get('lighthouse.validation_cache');
+
+        if (! isset($cacheConfig['enable']) || ! $cacheConfig['enable']) {
+            return DocumentValidator::validate($schema, $query, $validationRules);
+        }
+
+        $cacheKey = "lighthouse:validation:{$schemaHash}:{$queryHash}";
+
+        $cacheFactory = Container::getInstance()->make(CacheFactory::class);
+        assert($cacheFactory instanceof CacheFactory);
+
+        $store = $cacheFactory->store($cacheConfig['store']);
+        if ($store->get($cacheKey) === self::CACHEABLE_RULES_ERROR_FREE_RESULT) {
+            return [];
+        }
+
+        $result = DocumentValidator::validate($schema, $query, $validationRules);
+        if ($result !== []) {
+            return $result;
+        }
+
+        $store->put($cacheKey, self::CACHEABLE_RULES_ERROR_FREE_RESULT, $cacheConfig['ttl']);
+
+        return [];
```

Comment:
In this case we will have to serialize errors. I'm not sure all errors can be properly serialized. 

`GraphQL\Error\Error` does implement the `JsonSerializable` interface, but it doesn't mean it would be properly serialized/unserialized during cache handling.

Reply:
Thanks, I added an explanatory comment. Maybe we can improve this in the future.